### PR TITLE
Add check in pytest to enforce that tests have required markers

### LINKED ISSUE #170: Revisit functionality to enforce all tests to be categorized with a pytest.mark
See also IDAES/idaes-dev#1038.

Adding @dangunter and @andrewlee94 as "persons of interest".

### PR DESCRIPTION
**Pull request by lbianchi-lbl**
## Summary/Motivation
All tests in the pytest test suite should have at least one of the required markers (`unit`, `component`, `integration`).
This PR implements this by adding a check that runs when the tests items are collected and causes any test without a required marker to fail.

## Changes proposed in this PR:
- Add function in `conftest.py` to check that test items have required markers
- Refactor existing code running during `pytest_runtest_setup` for consistency

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Review by andrewlee94 — APPROVED** [review]
Looks good to me. Even better that it is failing because some tests are missing marks.

File: idaes/conftest.py [hunk 1]
```diff
@@ -29,37 +29,72 @@
...
+@pytest.hookimpl
+def pytest_runtest_setup(item):
+    _validate_required_markers(
+        item,
+        required_markers={"unit", "component", "integration"},
+        expected_count=1
+    )
+    _skip_for_unsupported_platforms(
+        item,
+        all_platforms={"darwin", "linux", "win32"},
```
Comment (dangunter) [inline idaes/conftest.py:43]: I prefer to define things like this in constants so you can change them without digging through the code.

File: idaes/conftest.py [hunk 2]
```diff
@@ -29,37 +29,72 @@
...
+def _validate_required_markers(item, required_markers=None, expected_count=1):
+    required_markers = set(required_markers or [])
+    item_markers = {marker.name for marker in item.iter_markers()}
+    required_markers_on_item = item_markers & required_markers
+    required_count = len(required_markers_on_item)
+    reason_to_fail = None
+    if required_count < expected_count:
+        reason_to_fail = 'Too few required markers'
+    if required_count > expected_count:
+        reason = 'Too many required markers'
```
Comment (dangunter) [inline idaes/conftest.py:93]: I think this should be `reason_to_fail`, as well

File: idaes/conftest.py [hunk 3]
```diff
@@ -29,37 +29,72 @@
...
+    if reason_to_fail:
+        extra_info = f'Expected: {expected_count} of {required_markers}; found: {item_markers}'
+        msg = f'{reason_to_fail} for test function "{item.name}". {extra_info}'
```
Comment (dangunter) [inline idaes/conftest.py:96]: Total nitpick, but appending extra_info to the string `msg` (multi-line) would be simpler than the indirection of defining a variable only to use it once, in the next line

**Review by dangunter — CHANGES_REQUESTED** [review]
One minor bug and a comment about the constants.

File: idaes/conftest.py (hunk 1 above)
Comment (lbianchi-lbl) [reply idaes/conftest.py:43]: Makes sense. I'll move it to a top-level module constant.

File: idaes/conftest.py (hunk 2 above)
Comment (lbianchi-lbl) [reply idaes/conftest.py:93]: Good catch, thanks! I'll change it to `reason_to_fail`.

File: idaes/conftest.py (hunk 3 above)
Comment (lbianchi-lbl) [reply idaes/conftest.py:96]: That's a good point, it'd look a bit cleaner. I'll change it following your suggestion.

**Review by ksbeattie — APPROVED** [review]
LGTM

**Review by dangunter — APPROVED** [review]
Looks great

**Review by andrewlee94 — APPROVED** [review]
Looks OK to me.

**Pull request merged**